specreduce/tests/test_extract.py

```diff
@@ -77,3 +78,28 @@ def test_array_trace():
 
     spectrum = boxcar(image, trace)
     assert np.allclose(spectrum.flux.value, np.full_like(spectrum.flux.value, 75.))
+
+
+def test_horne_variance_errors():
+    extract = HorneExtract()
+    trace = FlatTrace(image, 3.0)
+
+    # all zeros are treated as non-weighted (give non-zero fluxes)
+    err = np.zeros_like(image)
+    mask = np.zeros_like(image)
+    ext = extract(image.data, trace, variance=err, mask=mask, unit=u.Jy)
+    assert not np.all(ext == 0)
+
+    # single zero value adjusts mask (does not raise error)
+    err = np.ones_like(image)
+    err[0] = 0
+    mask = np.zeros_like(image)
+    ext = extract(image.data, trace, variance=err, mask=mask, unit=u.Jy)
+    assert not np.all(ext == 0)
+
+    # single negative value raises error
+    err = np.ones_like(image)
+    err[0] = -1
+    mask = np.zeros_like(image)
+    with pytest.raises(ValueError, match='variance must be fully positive'):
```

Comment:
Hinging the test on the exact phrase used in the traceback may be dangerous if it's edited innocuously in the future. I assume this scenario would then always pass, even if the `ValueError` was triggered for the reason being tested. Should there be a comment by the phrase in `extract.py`, or is there a more canonical way to address this concern?

Reply:
this line _requires_ the indented block to raise a value error with that exact message or else the test will fail.  So if we change the error message at any point, we will also need to update the test (or else the test will fail).  Is that what you meant or is there another situation to worry about?